### DISCUSSION TITLE
Help with dynamic outputs and chains of ops

### DISCUSSION THREAD

**Discussion by snoopfab** [dagster-io/dagster#5868] [Q&A] [answered]
I'm currently evaluating dagster for workflows orchestration.

I saw the map reduce example in the [documentation ](https://docs.dagster.io/concepts/ops-jobs-graphs/jobs-graphs#dynamic-mapping--collect).
However in my case i would like to map a chain of ops and not a single one

Example pseudo code: 

```
@op(out=DynamicOut(str))
def listTestCases(context):    
    for test_case in range(10) :
        testcase = f"testcase_{test_case}"
        yield DynamicOutput(
            value=testcase,
            # create a mapping key from the file name
            mapping_key=testcase,
        )


@op 
def A(context,testcase) : 
   context.log.info(f"[A] : {testcase}")

@op 
def B(context,testcase) : 
   context.log.info(f"[B] : {testcase}")

@op 
def C(context,testcase) : 
   context.log.info(f"[C] : {testcase}")

@job
def dynamic_job():
    listTestCases().map( A -> B -> C )

```

Expected results : 

```
                        A(testcase0)    ->  B(testcase0)   -> C(testcase0)
listTestCases   ->      A(testcase1)    ->  B(testcase1)   -> C(testcase1)
                            ...                 ...               ...
                        A(testcase9)    ->  B(testcase9)   -> C(testcase9)

```
where all the A -> B -C chains are executed in parrallel and listTestCases returns a dynamic results at runtime.

I can do that with a static range of test cases known at compile time but i did'nt find a way to do this with a number of test cases only known at runtime.

Is this possible with dagster ?

Thanks in advance for your time and help !

**Comment by sryza** [comment, marked as answer]
Hi @snoopfab, the following should work:
```
@job
def dynamic_job():
    listTestCases().map(A).map(B).map(C)
```

**Comment by snoopfab** [comment]
Thanks for your proposition.

I forgot to mention that i need the whole process to be **asynchronous** and **parallel**.

What i want is the B1 task to start as soon as the A1 task is completed and the C1 task as soon as the B1 is completed.

In your proposed solution , All the A_i tasks are executed in parallel, then the results are collected, and after this All the B_i tasks are executed etc ... 

I can do what i want (asyncronicity and parallelism) with a static number of testcase known at job definition time but in my case it can only be known at runtime.

Moreover, i cannot aggregate the A, B,C tasks in only one task as the B_i tasks needs a specific celery worker with unique ressources (ex:  A GPU).

**Reply by alangenfeld** [reply]
> In your proposed solution , All the A_i tasks are executed in parallel, then the results are collected, and after this All the B_i tasks are executed etc ...

The example above does not have calls to `collect` so should achieve the goal you describe. 

The `@job` decorated function is used to create the shape of the dag, it is executed only at init time to build a dependency structure which dagster then drives execution over. It is not executed or used at runtime, so intuition about how that python code would behave can be misleading.


![Screen Shot 2021-12-09 at 10 23 55 AM](https://user-images.githubusercontent.com/202219/145435357-4d49824f-4b80-456e-9300-85bfb9a2d7a7.png)

**Reply by snoopfab** [reply]
Thank you very much for the cool graphs  !

I was  running the example with a process executor locally hence the tasks were executed one by one so i thought it was not working as expected.
But when i tried with three properly configured celery workers  (concurrency level=3) the results is now as expected

Thanks again for your time !

![dagit](https://user-images.githubusercontent.com/22932323/145445871-6efe23ca-75f9-42d4-b209-7b73c356e2b5.png)
